### PR TITLE
Implement tdi_port_manager.cc

### PR DESCRIPTION
**Pull request by ffoulkes**
- Converted TdiPortManager from an abstract class to a base class.
- Moved common definitions from subclasses to the base class.
- Eliminated vestigial Es2kPortManager methods.

### CODE REVIEW

File: stratum/hal/lib/tdi/tdi_port_manager.h [hunk 1]
```diff
@@ -70,6 +71,26 @@ class TdiPortManager {
...
+  // RW mutex lock for protecting the singleton instance initialization and
+  // reading it back from other threads. Unlike other singleton classes, we
+  // use RW lock as we need the pointer to class to be returned.
+  static absl::Mutex init_lock_;
+
+  // Timeout for Write() operations on port status events.
+  static constexpr absl::Duration kWriteTimeout = absl::InfiniteDuration();
+
+  // Default MTU for ports on ES2K.
+  static constexpr int32 kBfDefaultMtu = 10 * 1024;  // 10K
```
Comment (5abeel) [inline stratum/hal/lib/tdi/tdi_port_manager.h]: Hopefully these dont need to be change based on different targets. But we could possibly override in future in derived classes if needed.

File: stratum/hal/lib/tdi/dpdk/dpdk_port_manager.cc [hunk 2]
```diff
@@ -86,19 +80,6 @@ ::util::Status DpdkPortManager::GetPortCounters(int device, int port,
   return ::util::OkStatus();
 }
 
-::util::Status DpdkPortManager::RegisterPortStatusEventWriter(
```
Comment (5abeel) [inline stratum/hal/lib/tdi/dpdk/dpdk_port_manager.cc:89]: Do the relevent MOCK routines need to be cleaned up in the dpdk_port_manager_mock.h?

Comment (5abeel) [reply stratum/hal/lib/tdi/dpdk/dpdk_port_manager.cc:89]: Same comment for es2k and tofino targets

**Review by 5abeel — APPROVED** [review]
Couple of comments - please verify if more cleanup needed in mock files. Rest LGTM

File: stratum/hal/lib/tdi/dpdk/dpdk_port_manager.cc (hunk 2 above)
Comment (ffoulkes) [reply stratum/hal/lib/tdi/dpdk/dpdk_port_manager.cc:89]: I don't think so. `Es2kPortManagerMock` extends `Es2kPortManager`. It mocks all the methods the parent class supports, including the ones it inherits from `TdiPortManager`. The same is true of the other PortManagerMocks. I believe this is the correct way to define the mock.

File: stratum/hal/lib/tdi/tdi_port_manager.h (hunk 1 above)
Comment (ffoulkes) [reply stratum/hal/lib/tdi/tdi_port_manager.h]: Good catch. On closer examination, it turns out that the only target that uses this constant is `tofino`.

I've moved the definition from `TdiPortManager` to `TofinoPortManager`.

**Review by nupuruttarwar — APPROVED** [review]
LGTM

**Pull request merged**